src/parser/nmodl_driver.cpp

```diff
@@ -24,60 +24,122 @@ std::shared_ptr<ast::Program> NmodlDriver::parse_stream(std::istream& in) {
     NmodlLexer scanner(*this, &in);
     NmodlParser parser(scanner, *this);
 
-    this->lexer = &scanner;
-    this->parser = &parser;
-
     scanner.set_debug(trace_scanner);
     parser.set_debug_level(trace_parser);
     parser.parse();
     return astRoot;
 }
 
-//// parse nmodl file
-std::shared_ptr<ast::Program> NmodlDriver::parse_file(const std::string& filename) {
+std::shared_ptr<ast::Program> NmodlDriver::parse_file(const std::string& filename,
+                                                      const location* loc) {
     std::ifstream in(filename.c_str());
     stream_name = filename;
 
+    {
+        const auto last_slash = filename.find_last_of(utils::pathsep);
+        if (utils::file_is_abs(filename)) {
+            const auto path_prefix = filename.substr(0, last_slash + 1);
+            library.push_current_directory(path_prefix);
+        } else if (last_slash == std::string::npos) {
+            library.push_current_directory(utils::cwd());
+        } else {
+            const auto path_prefix = filename.substr(0, last_slash + 1);
+            const auto path = utils::cwd() + utils::pathsep + path_prefix;
+            library.push_current_directory(path);
+        }
+    }
+
     if (!in.good()) {
-        logger->error("Can not open file : {}", filename);
-        return nullptr;
+        std::ostringstream oss;
+        if (loc == nullptr) {
+            oss << "NMODL Parser Error : ";
+        }
+        oss << "can not open file : " << filename;
+        if (loc != nullptr) {
+            parse_error(*loc, oss.str());
+        } else {
+            throw std::runtime_error(oss.str());
+        }
     }
     parse_stream(in);
+    library.pop_current_directory();
     return astRoot;
 }
 
-/// parser nmodl provided as string (used for testing)
 std::shared_ptr<ast::Program> NmodlDriver::parse_string(const std::string& input) {
     std::istringstream iss(input);
     parse_stream(iss);
     return astRoot;
 }
 
-void NmodlDriver::error(const std::string& m, const class location& l) {
-    std::cerr << l << " : " << m << std::endl;
-}
+std::shared_ptr<ast::Include> NmodlDriver::parse_include(const std::string& name,
+                                                         const location& loc) {
+    // Try to find directory containing the file to import
+    const auto directory_path = library.find_file(name);
+
+    // Complete path of file (directory + filename).
+    std::string absolute_path = name;
+
+    if (!directory_path.empty()) {
+        absolute_path = directory_path + std::string(1, utils::pathsep) + name;
+    }
+
+    // Detect recursive inclusion.
+    if (open_files.find(absolute_path) != open_files.end()) {
+        std::ostringstream oss;
+        oss << name << ": recursive inclusion.\n"
```

Comment:
nitpick: use .format here?

Reply:
The error message is constructed with a `nmodl::parser::location` that can work with `std::basic_ostream` but not the _format_ library. I wanted to extend the format library to deal with `nmodl::parser::location` as instructed [here](https://fmt.dev/latest/api.html#formatting-user-defined-types) but realized that every bison parser in NMODL has its owned `location.hpp` file:
* `${CMAKE_BINARY_DIR}/src/parser/diffeq/location.hh`
* `${CMAKE_BINARY_DIR}/src/parser/nmodl/location.hh`
* `${CMAKE_BINARY_DIR}/src/parser/unit/location.hh`

Each of them defines a `location` class, all in the same `nmodl::parser` namespace. There should be either one single `nmodl::parser::location` class used by all parsers or they should all be declared in a dedicated namespace. This is an issue that should be addressed before considering extending the _format_ to deal with `nmodl::parser::location`.

I will create a dedicated issue for this so that we can discuss

Reply:
thx! Interesting...